src/shared/reboot-util.c

```diff
@@ -107,3 +108,43 @@ int shall_restore_state(void) {
 
         return r > 0 ? ret : true;
 }
+
+static int xen_kexec_loaded(void) {
+#if HAVE_XENCTRL
+        xc_interface *xch;
+        int r;
+
+        r = dlopen_xenctrl();
+        if (r < 0)
+                return r;
+
+        if (access("/proc/xen", F_OK) < 0)
+                return log_debug_errno(errno, "Cannot access /proc/xen: %m");
+
+        xch = sym_xc_interface_open(NULL, NULL, 0);
+        if (!xch)
+                return log_debug_errno(errno, "Failed to open Xen control interface: %m");
```

Comment:
hmm, are you sure xc_interface_open() sets errno? I can't find any docs, and google for code snippets shows noone using errno afterwards?

Reply:
> but it doesn't look like the implementation does:
>
> https://sources.debian.org/src/xen/4.16.2-2/tools/libs/ctrl/xc_private.c/?hl=27#L74

In the err cases, errno has been set by previous functions calls.